### PR TITLE
CNS-802-make-provider-commission-increase-take-longer-time

### PR DESCRIPTION
**Pull request by Yaroms**

### CODE REVIEW

File: x/pairing/keeper/staking.go [hunk 1]
```diff
@@ -92,12 +95,31 @@ func (k Keeper) StakeNewEntry(ctx sdk.Context, validator, creator, chainID strin
 		}
 		details = append(details, utils.Attribute{Key: "moniker", Value: moniker})
 
+		if !existingEntry.DelegateTotal.IsZero() {
+			diff := ctx.BlockTime().UTC().Unix() - int64(existingEntry.LastChange)
+			_ = diff
```
Comment (oren-lava) [inline x/pairing/keeper/staking.go:100]: remove

File: x/pairing/keeper/staking.go [hunk 2]
```diff
@@ -92,12 +95,31 @@ func (k Keeper) StakeNewEntry(ctx sdk.Context, validator, creator, chainID strin
...
+			if ctx.BlockTime().UTC().Unix()-int64(existingEntry.LastChange) < int64((time.Hour * 24).Seconds()) {
```
Comment (oren-lava) [inline x/pairing/keeper/staking.go:101]: why are you not using `diff`?
also, make the 24H a constant

File: x/pairing/keeper/staking.go [hunk 3]
```diff
@@ -92,12 +95,31 @@ func (k Keeper) StakeNewEntry(ctx sdk.Context, validator, creator, chainID strin
 		}
 		details = append(details, utils.Attribute{Key: "moniker", Value: moniker})
 
+		if !existingEntry.DelegateTotal.IsZero() {
```
Comment (oren-lava) [inline x/pairing/keeper/staking.go:98]: add a comment that summarizes all the new limitations. Also, update the README

File: x/pairing/keeper/staking.go [hunk 4]
```diff
@@ -92,12 +95,31 @@ func (k Keeper) StakeNewEntry(ctx sdk.Context, validator, creator, chainID strin
...
+			diff := ctx.BlockTime().UTC().Unix() - int64(existingEntry.LastChange)
+			_ = diff
+			if ctx.BlockTime().UTC().Unix()-int64(existingEntry.LastChange) < int64((time.Hour * 24).Seconds()) {
+				if delegationCommission != existingEntry.DelegateCommission || existingEntry.DelegateLimit != delegationLimit {
+					return utils.LavaFormatWarning("stake entry commmision or delegate limit can only be changes once in 24H", nil)
```
Comment (oren-lava) [inline x/pairing/keeper/staking.go:103]: change the error description to use the 24H constant (and not write literally "24H")

File: x/pairing/keeper/staking.go [hunk 5]
```diff
@@ -12,6 +13,8 @@ import (
 	spectypes "github.com/lavanet/lava/x/spec/types"
 )
 
+const MAX_CHANGE_RATE = 1
```
Comment (oren-lava) [inline x/pairing/keeper/staking.go:16]: why is this here and not in the types folder somewhere (we usually put constants in types.go)

File: x/pairing/keeper/msg_server_stake_provider_test.go [hunk 6]
```diff
@@ -812,3 +813,47 @@ func TestUnfreezeWithDelegations(t *testing.T) {
 	_, err = ts.TxPairingUnfreezeProvider(provider, ts.spec.Index)
 	require.NoError(t, err)
 }
+
+func TestCommisionChange(t *testing.T) {
```
Comment (oren-lava) [inline x/pairing/keeper/msg_server_stake_provider_test.go:817]: add a comment that sunnarizes all the scenarios you're testing (you can also add comments inside the function, some of the checks are unexplained)

File: x/pairing/keeper/msg_server_stake_provider_test.go [hunk 7]
```diff
@@ -812,3 +813,47 @@ func TestUnfreezeWithDelegations(t *testing.T) {
...
+
+	_, err = ts.TxPairingStakeProviderFull(provider, ts.spec.Index, ts.spec.MinStakeProvider, nil, 0, "", 60, 140)
+	require.NoError(t, err)
+
+	// add delegator and delegate to provider
+	_, consumer := ts.AddAccount(common.CONSUMER, 1, testBalance)
+	_, err = ts.TxDualstakingDelegate(consumer, provider, ts.spec.Index, ts.spec.MinStakeProvider)
+	require.NoError(t, err)
+	ts.AdvanceEpoch() // apply delegation
+	ts.AdvanceBlock(time.Hour * 25)
```
Comment (oren-lava) [inline x/pairing/keeper/msg_server_stake_provider_test.go:841]: specify you advance 25H to be able to change the delegation limit/commission

**Review by oren-lava — CHANGES_REQUESTED** [review]

File: x/pairing/keeper/staking.go (hunk 5 above)
Comment (Yaroms) [reply x/pairing/keeper/staking.go:16]: these are only related to this file

Comment (Yaroms) [reply x/pairing/keeper/staking.go]: https://github.com/lavanet/lava/pull/1263/commits/5cda7df3147294fb3b1aaa960ff157f5d8e2b00b

File: x/pairing/keeper/staking.go (hunk 4 above)
Comment (Yaroms) [reply x/pairing/keeper/staking.go]: https://github.com/lavanet/lava/pull/1263/commits/5cda7df3147294fb3b1aaa960ff157f5d8e2b00b

File: x/pairing/keeper/staking.go (hunk 2 above)
Comment (Yaroms) [reply x/pairing/keeper/staking.go]: https://github.com/lavanet/lava/pull/1263/commits/5cda7df3147294fb3b1aaa960ff157f5d8e2b00b

File: x/pairing/keeper/staking.go (hunk 1 above)
Comment (Yaroms) [reply x/pairing/keeper/staking.go]: https://github.com/lavanet/lava/pull/1263/commits/5cda7df3147294fb3b1aaa960ff157f5d8e2b00b

File: x/pairing/keeper/staking.go [hunk 8]
```diff
@@ -92,12 +98,33 @@ func (k Keeper) StakeNewEntry(ctx sdk.Context, validator, creator, chainID strin
...
+			// if there was a change in the last 24h than we dont allow changes
+			if ctx.BlockTime().UTC().Unix()-int64(existingEntry.LastChange) < int64(CHANGE_WINDOW.Seconds()) {
+				if delegationCommission != existingEntry.DelegateCommission || existingEntry.DelegateLimit != delegationLimit {
+					return utils.LavaFormatWarning(fmt.Sprintf("stake entry commmision or delegate limit can only be changes once in %s", CHANGE_WINDOW), nil)
+				}
+			}
+
+			// check that the change is not mode than MAX_CHANGE_RATE
+			if int64(delegationCommission)-int64(existingEntry.DelegateCommission) > MAX_CHANGE_RATE {
+				return utils.LavaFormatWarning("stake entry commission increase too high", fmt.Errorf("commission change cannot increase by more than %d at a time", MAX_CHANGE_RATE))
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/pairing/keeper/staking.go:112]: done

File: x/pairing/keeper/staking.go [hunk 9]
```diff
@@ -92,12 +98,33 @@ func (k Keeper) StakeNewEntry(ctx sdk.Context, validator, creator, chainID strin
...
+			}
+
+			// check that the change in delegation limit is decreasing and that new_limit*100/old_limit < (100-MAX_CHANGE_RATE)
+			if delegationLimit.IsLT(existingEntry.DelegateLimit) && delegationLimit.Amount.MulRaw(100).Quo(existingEntry.DelegateLimit.Amount).LT(sdk.NewInt(100-MAX_CHANGE_RATE)) {
+				return utils.LavaFormatWarning("stake entry DelegateLimit decrease too high", fmt.Errorf("DelegateLimit change cannot decrease by more than %d at a time", MAX_CHANGE_RATE))
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/pairing/keeper/staking.go:117]: done

File: x/pairing/keeper/staking.go [hunk 10]
```diff
@@ -92,12 +98,33 @@ func (k Keeper) StakeNewEntry(ctx sdk.Context, validator, creator, chainID strin
 		}
 		details = append(details, utils.Attribute{Key: "moniker", Value: moniker})
 
+		// if the provider has no delegations then we dont limit the changes
+		if !existingEntry.DelegateTotal.IsZero() {
+			// if there was a change in the last 24h than we dont allow changes
+			if ctx.BlockTime().UTC().Unix()-int64(existingEntry.LastChange) < int64(CHANGE_WINDOW.Seconds()) {
+				if delegationCommission != existingEntry.DelegateCommission || existingEntry.DelegateLimit != delegationLimit {
+					return utils.LavaFormatWarning(fmt.Sprintf("stake entry commmision or delegate limit can only be changes once in %s", CHANGE_WINDOW), nil)
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/pairing/keeper/staking.go:106]: done

**Review by omerlavanet — APPROVED** [review]
